crowbar_framework/lib/crowbar/lock/shared_non_blocking.rb

```diff
@@ -0,0 +1,141 @@
+        end
+      end
+      @executor = options.fetch :executor, default_executor
+
+      # To implement shared locking semantics, we keep track of all
+      # "readers" via a 'foo.lock.d/' directory which contains one
+      # file per "reader":
+      @readers_dir = path + ".d"
+
+      # In order to provide a lock for the writer, we also ensure that
+      # a 'foo.lock' file exists if and only if the 'foo.lock.d/'
+      # directory contains one or more readers.  Doing this safely
+      # requires a "meta-lock" in order to avoid races when we create
+      # or remove the 'foo.lock' file.  We implement that meta-lock
+      # using flock(1) on another 'foo.lock.meta' file.
+      @meta_lock = path + ".meta"
+    end
+
+    def acquire
+      if locked?
+        logger.warn("#{owner} tried to re-acquire remote lock #{path} on #{node}")
+        return
+      end
+
+      logger.debug("#{owner} acquiring remote lock #{path} on #{node}")
+      shell = <<-EOSHELL.strip
+        mkdir -p "#{@readers_dir}"
+        file="#{@readers_dir}/#{owner}"
+        if [ -e "$file" ]; then
+          echo "ERROR: #{owner} already had lock via $file" >&2
+          exit 1
+        fi
+        echo "#{reason}" > "$file"
+        echo "`date` acquired by #{owner}" >> "#{path}"
+      EOSHELL
+      run_with_exclusive_lock(shell)
+      logger.debug("#{owner} acquired remote lock #{path} on #{node}")
+      @locked = true
+      self
+    end
+
+    def release
+      logger.debug("#{owner} releasing remote lock #{path} on #{node}")
+      shell = <<-EOSHELL.strip
+        rm -f #{@readers_dir}/#{owner}
+        if [ -d #{@readers_dir} ] && ! ls #{@readers_dir} | grep -q .; then
+          rm -f #{path}
+        else
+          echo "`date` released by #{owner}" >> "#{path}"
+        fi
+      EOSHELL
+      run_with_exclusive_lock(shell)
+      logger.debug("#{owner} released remote lock #{path} on #{node}")
+      @locked = false
+      self
+    end
+
+    # Use this when there are no more readers.  Mainly useful for
+    # ensuring the test suite doesn't leave temporary files around.
+    def clean
+      raise "Can't clean active lock I hold" if locked?
+
+      logger.debug("#{owner} cleaning remote lock #{path} on #{node}")
+      shell = <<-EOSHELL.strip
+        if [ -d #{@readers_dir} ] && ls #{@readers_dir} | grep -q .; then
+          echo "Cannot clean active lock someone else holds" >&2
+          exit 1
+        fi
+        if [ -d #{@readers_dir} ]; then
+          rmdir #{@readers_dir}
+        fi
+        rm -f #{@meta_lock}
+      EOSHELL
+      run_with_exclusive_lock(shell)
+    end
+
+    private
+
+    # Run a critical region with an exclusive lock so that no
+    # concurrent thread can race with it.
+    def run_with_exclusive_lock(shell)
+      cmd = "flock #{@meta_lock} bash -es"
+      logger.debug("Will execute on node #{node}: #{cmd}" + \
+                   "and feed the following to STDIN:\n#{shell}")
```

Comment:
Do we really want to put the shell command there? Don't we know what it will be anyway?